docs/proposals/failover/application-failover.md

```diff
@@ -0,0 +1,373 @@
+type ApplicationFailoverBehavior struct {
+    // InitiateGates indicates the preconditions of the failover process.
+    // If specified, all gates will be evaluated for prerequisites.
+    // +optional
+    InitiateGates []FailoverInitiateGate `json:"initiateGates,omitempty"`
+  
+    // ActiveGates indicates the conditions of performing the failover process.
+    //
+    // +kubebuilder:validation:MinItems=1
+    // +required
+    ActiveGates []FailoverActiveGate `json:"activeGates"`
+  
+    // TolerationSeconds represents the period of time the Karmada should wait
+    // before performing failover. Defaults to 10s.
+    //
+    // +kubebuilder:default=10
+    // +optional
+    TolerationSeconds int32
+  
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Immediately", "Graciously" and "Never".
+    // Defaults to "Graciously".
+    //
+    // +kubebuilder:validation:Enum=Immediately;Graciously;Never
+    // +optional
+    PurgeMode string `json:"purgeMode,omitempty"`
+  
+    // EscapeSeconds represents the period of time the cluster from which the
+    // application was migrated from should be escaped.
+    // Defaults to 600s.
+    //
+    // +kubebuilder:default=600
+    // +optional
+    EscapeSeconds int32 `json:"escapeSeconds,omitempty"`
+}
+
+// FailoverInitiateGate represents the preconditions of the failover process.
+type FailoverInitiateGate struct {
+    // ConditionType refers to a condition in the object's condition list with
+    // matching type. If the status of the specified condition is "true" means pass.
+    // +optional
+    ConditionType string `json:"conditionType,omitempty"`
+  
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Unhealthy" and "Unknown".
+    // +optional
+    HealthyState string `json:"healthyState,omitempty"`
+}
+
+// FailoverActiveGate represents the conditions of performing the failover process.
+type FailoverActiveGate struct {
+    // ConditionType refers to a condition in the object's condition list with
+    // matching type. If the status of the specified condition is "true" means pass.
+    // +optional
+    ConditionType string `json:"conditionType,omitempty"`
+  
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Unhealthy" and "Unknown".
+    // +optional
+    HealthyState string `json:"healthyState,omitempty"`
+}
+```
+
+For example:
+
+```yaml
+apiVersion: policy.karmada.io/v1alpha1
+kind: PropagationPolicy
+metadata:
+  name: nginx-propagation
+spec:
+  resourceSelectors:
+    - apiVersion: apps/v1
+      kind: Deployment
+      name: nginx
+    - apiVersion: apps/v1
+      kind: StatefulSet
+      name: mysql
+  failover:
+    - apiVersion: apps/v1
+      kind: Deployment
```

Comment:
defined here `.spec.failover[0]. apiVersion` and `.spec.failover[0]. kind` mean?

What happens if I have two pp and apiVersion and kind are set to the same for both pp, but `behavior.application.activeGates` is not the same?

Reply:
Sorry for the slow reply. For these two pp, one resource will only apply one of above policies. So it will use the failover behavior in the hit policy.

Reply:
Personality, I don't like to introduce an extra layer to declare the `apiVersion/kind` that the failover rules applied to.

My concern is for most cases, there is usually only one application in the PropagatioPolicy, but this design forces people to declare the `apiVersion/kind` one more time, a little bit redundant.

But, on the other hand, we might need to support declaring different behaviors for different applications in the same PropagationPolicy. 

I don't have a better idea yet.

Reply:
> (1) When users set different applications in the same pp and config the behavior by condition, they may need to live with the configuration of one of the applications not taking effect.

I probably don't buy it if I'm the user.